src/riak_kv_quanta.erl

```diff
@@ -0,0 +1,247 @@
+-module(riak_kv_quanta).
+
+-export([quanta/3,
+         timestamp_to_ms/1,
+         ms_to_timestamp/1]).
+
+-type time_ms() :: non_neg_integer().
+-type time_unit() :: y | mo | d | h | m | s.
+-type err() :: {error, term()}.
+
+%% The Number of Days from Jan 1, 0 to Jan 1, 1970
+%% We need this to compute years and months properly including leap years and variable length
+%% months.
+-define(DAYS_FROM_0_TO_1970, 719528).
+
+-ifdef(TEST).
+-ifdef(EQC).
+-include_lib("eqc/include/eqc.hrl").
+-define(QC_OUT(P),
+        eqc:on_output(fun(Str, Args) ->
+                              io:format(user, Str, Args) end, P)).
+-compile(export_all).
+-endif.
+-endif.
+
+%% @doc Return the time in milliseconds since 00:00 GMT Jan 1, 1970 (Unix Epoch)
+-spec timestamp_to_ms(erlang:timestamp()) -> time_ms().
+timestamp_to_ms({Mega, Secs, Micro}) ->
+    Mega*1000000000 + Secs*1000 + Micro div 1000.
+
+-spec ms_to_timestamp(time_ms()) -> erlang:timestamp().
+ms_to_timestamp(Time) ->
+    Seconds = Time div 1000,
+    MicroSeconds = (Time rem 1000) * 1000,
+    {0, Seconds, MicroSeconds}.
+
+-spec quanta(time_ms(), non_neg_integer(), time_unit()) -> time_ms() | err().
+quanta(Time, QuantaSize, Unit) ->
+    case lists:member(Unit, [d, h, m, s]) of
+        true ->
+            Ms = unit_to_ms(Unit),
+            Diff = Time rem (QuantaSize*Ms),
+            Time - Diff;
+        false ->
+            large_quanta(Time, QuantaSize, Unit)
+    end.
+
+-spec large_quanta(time_ms(), non_neg_integer(), mo | y) -> time_ms() | err().
+large_quanta(Time, QuantaSize, mo) ->
+    Timestamp = ms_to_timestamp(Time),
+    Month = months_since_1970(Timestamp),
+    MonthQuanta = Month - (Month rem QuantaSize),
+    months_since_1970_to_ms(MonthQuanta);
+large_quanta(Time, QuantaSize, y) ->
+    Timestamp = ms_to_timestamp(Time),
+    {{Year, _, _}, _} = calendar:now_to_universal_time(Timestamp),
+    YearsSince1970 = Year - 1970,
+    YearQuanta = Year - (YearsSince1970 rem QuantaSize),
+    years_since_1970_to_ms(YearQuanta);
+large_quanta(_, _, Unit) ->
+    {error, {invalid_unit, Unit}}.
+
```

Comment:
I like the second way. You could also collapse the guard with a macro, e.g.

```erlang
-define(IS_SMALL_UNIT(U), U == d orelse U == h orelse U == m orelse U == s).

quanta(Time, QuantaSize, Unit) when ?IS_SMALL_UNIT(Unit) ->
   %...
```

I'd probably only do that though if you were going to reuse the guard (or as part of some other boolean expression). Note `=:=` is only necessary for numbers.